File: Rentoza/src/main/java/org/example/rentoza/car/CarService.java
package org.example.rentoza.car;

import org.example.rentoza.booking.BookingRepository;
import org.example.rentoza.booking.BookingStatus;
import org.example.rentoza.car.dto.CarRequestDTO;
import org.example.rentoza.car.dto.CarResponseDTO;
import org.example.rentoza.car.dto.CarSearchCriteria;
import org.example.rentoza.car.storage.CarImageStorageService;
import org.example.rentoza.common.GeoPoint;
import org.example.rentoza.exception.ResourceNotFoundException;
import org.example.rentoza.review.ReviewDirection;
import org.example.rentoza.review.ReviewRepository;
import org.example.rentoza.user.User;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.stream.Collectors;

import org.springframework.web.multipart.MultipartFile;

@Service
public class CarService {

    private static final Logger log = LoggerFactory.getLogger(CarService.class);

    private final CarRepository repo;
    private final BookingRepository bookingRepo;
    private final ReviewRepository reviewRepo;
    private final org.example.rentoza.security.CurrentUser currentUser;
    private final CarImageStorageService carImageStorageService;

    public CarService(
            CarRepository repo,
            BookingRepository bookingRepo,
            ReviewRepository reviewRepo,
            org.example.rentoza.security.CurrentUser currentUser,
            CarImageStorageService carImageStorageService
    ) {
        this.repo = repo;
        this.bookingRepo = bookingRepo;
        this.reviewRepo = reviewRepo;
        this.currentUser = currentUser;
        this.carImageStorageService = carImageStorageService;
    }

    @Transactional
    public Car addCarWithLocalImages(CarRequestDTO dto, User owner, List<MultipartFile> images) {
        // This path is used by multipart/form-data uploads.
        // We intentionally DO NOT persist any base64 data into the database.
        dto.setImageUrl(null);
        dto.setImageUrls(null);

        Car savedCar = addCar(dto, owner);

        List<String> imageUrls = carImageStorageService.storeCarImages(savedCar.getId(), images);
        if (!imageUrls.isEmpty()) {
            savedCar.setImageUrl(imageUrls.get(0));
            savedCar.setImageUrls(new ArrayList<>(imageUrls));
            savedCar = repo.save(savedCar);
        }

        return savedCar;
    }

    public Car addCar(CarRequestDTO dto, User owner) {
        // Validate required fields
        if (dto.getBrand() == null || dto.getBrand().isBlank()) {
            throw new RuntimeException("Brand is required");
        }
        if (dto.getModel() == null || dto.getModel().isBlank()) {
            throw new RuntimeException("Model is required");
        }
        // BigDecimal price validation (minimum 10 RSD)
        if (dto.getPricePerDay() == null || dto.getPricePerDay().compareTo(BigDecimal.TEN) < 0) {
            throw new RuntimeException("Price per day must be at least 10 RSD");
        }
        // Geospatial coordinates are REQUIRED (Phase 2.4)
        if (dto.getLocationLatitude() == null || dto.getLocationLongitude() == null) {
            throw new RuntimeException("Location coordinates (latitude/longitude) are required");
        }

        // Create new car entity with default values
        Car car = new Car();

        // Map basic required fields
        car.setBrand(dto.getBrand().trim());
        car.setModel(dto.getModel().trim());
        car.setYear(dto.getYear());
        car.setPricePerDay(dto.getPricePerDay());
        car.setLocation(dto.getLocation().trim().toLowerCase());
        car.setOwner(owner);
        car.setApprovalStatus(ApprovalStatus.PENDING);
        car.setAvailable(false);

        // Set geospatial location (Phase 2.4 - REQUIRED)
        GeoPoint geoPoint = new GeoPoint();
        geoPoint.setLatitude(dto.getLocationLatitude());
        geoPoint.setLongitude(dto.getLocationLongitude());
        geoPoint.setAddress(dto.getLocationAddress() != null ? dto.getLocationAddress() : dto.getLocation());
        geoPoint.setCity(dto.getLocationCity());
        geoPoint.setZipCode(dto.getLocationZipCode());
        car.setLocationGeoPoint(geoPoint);

        // Map optional license plate
        if (dto.getLicensePlate() != null && !dto.getLicensePlate().isBlank()) {
            car.setLicensePlate(dto.getLicensePlate().trim());
        }

        // Map optional description
        if (dto.getDescription() != null && !dto.getDescription().isBlank()) {
            car.setDescription(dto.getDescription().trim());
        }

        // Map specifications (override defaults if provided)
        if (dto.getSeats() != null) {
            car.setSeats(dto.getSeats());
        }
        if (dto.getFuelType() != null) {
            car.setFuelType(dto.getFuelType());
        }
        if (dto.getFuelConsumption() != null) {
            car.setFuelConsumption(dto.getFuelConsumption());
        }
        if (dto.getTransmissionType() != null) {
            car.setTransmissionType(dto.getTransmissionType());
        }

        // Map features (initialize empty set, then add if provided)
        if (dto.getFeatures() != null && !dto.getFeatures().isEmpty()) {
            car.setFeatures(new HashSet<>(dto.getFeatures()));
        }

        // Map add-ons (initialize empty set, then add if provided)
        if (dto.getAddOns() != null && !dto.getAddOns().isEmpty()) {
            car.setAddOns(new HashSet<>(dto.getAddOns()));
        }

        // Map rental policies (override defaults if provided)
        if (dto.getCancellationPolicy() != null) {
            car.setCancellationPolicy(dto.getCancellationPolicy());
        }
        if (dto.getMinRentalDays() != null) {
            car.setMinRentalDays(dto.getMinRentalDays());
        }
        if (dto.getMaxRentalDays() != null) {
            car.setMaxRentalDays(dto.getMaxRentalDays());
        }

        // Map images
        if (dto.getImageUrl() != null && !dto.getImageUrl().isBlank()) {
            car.setImageUrl(dto.getImageUrl());
        }
        if (dto.getImageUrls() != null && !dto.getImageUrls().isEmpty()) {
            car.setImageUrls(new ArrayList<>(dto.getImageUrls()));
        }

        Car savedCar = repo.save(car);

        // Log successful persistence for debugging
        log.debug("Car saved successfully: id={}, brand={}, model={}, year={}, seats={}, fuelType={}, " +
                  "transmissionType={}, features={}, addOns={}, cancellationPolicy={}, description={}",
                savedCar.getId(), savedCar.getBrand(), savedCar.getModel(), savedCar.getYear(),
                savedCar.getSeats(), savedCar.getFuelType(), savedCar.getTransmissionType(),
                savedCar.getFeatures().size(), savedCar.getAddOns().size(),
                savedCar.getCancellationPolicy(),
                savedCar.getDescription() != null ? savedCar.getDescription().substring(0, Math.min(50, savedCar.getDescription().length())) + "..." : "null");

        return savedCar;
    }

    @Transactional(readOnly = true)
    public List<CarResponseDTO> getAllCars() {
        // Public listing - only show available cars to users
        // Privacy: Use fuzzy locations for non-owners
        Long currentUserId = currentUser.idOrNull();
        return repo.findByAvailableTrue()
                .stream()
                .map(car -> mapToResponseWithPrivacy(car, currentUserId))
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public List<CarResponseDTO> getCarsByLocation(String location) {
        // Public listing - only show available cars to users
        // Privacy: Use fuzzy locations for non-owners
        Long currentUserId = currentUser.idOrNull();
        return repo.findByLocationIgnoreCaseAndAvailableTrue(location)
                .stream()
                .map(car -> mapToResponseWithPrivacy(car, currentUserId))
                .collect(Collectors.toList());
    }

    /**
     * Get cars by owner email with ownership verification.
     * RLS-ENFORCED: Returns cars only if requester is the owner or admin.
     * Prevents Owner A from viewing Owner B's private inventory.
     * 
     * @param email Owner's email
     * @return List of cars owned by the user
     * @throws org.springframework.security.access.AccessDeniedException if requester is not the owner or admin
     */
    @Transactional(readOnly = true)
    public List<CarResponseDTO> getCarsByOwner(String email) {
        // RLS ENFORCEMENT: Verify requester is the owner or admin
        String requesterEmail = currentUser.email();
        if (!requesterEmail.equalsIgnoreCase(email) && !currentUser.isAdmin()) {
            throw new org.springframework.security.access.AccessDeniedException(
                    "Unauthorized to access cars for owner: " + email
            );
        }
        
        Long currentUserId = currentUser.idOrNull();
        return repo.findByOwnerEmailIgnoreCase(email)
                .stream()
                .map(car -> mapToResponseWithPrivacy(car, currentUserId)) // Owners will see exact location
                .collect(Collectors.toList());
    }

    /**
     * Get car by ID with full details (features, addOns, images).
     * Uses @EntityGraph to load all collections in a single query.
     * 
     * PERFORMANCE:
     * - Single query with LEFT JOIN FETCH for all collections
     * - No N+1 queries when accessing features/addOns
     * - Optimized for detail page views
     * 
     * @throws ResourceNotFoundException if car not found
     */
    @Transactional(readOnly = true)
    public CarResponseDTO getCarById(Long id) {
        log.debug("[CarService] Fetching car with ID: {}", id);
        
        // Use the detail-loading method to prevent LazyInitializationException
        Car car = repo.findWithDetailsById(id)
                .orElseThrow(() -> {
                    log.warn("[CarService] Car not found with ID: {}", id);
                    return new ResourceNotFoundException("Car not found with ID: " + id);
                });

        log.debug("[CarService] Found car: {} {} (ID={})", car.getBrand(), car.getModel(), car.getId());

        // Privacy check: Is current user the owner or has an active booking?
        Long currentUserId = currentUser.idOrNull();
        boolean hasActiveBooking = false;
        
        if (currentUserId != null) {
            // Check for active booking (APPROVED, IN_TRIP, etc.)
            hasActiveBooking = bookingRepo.findByCarIdAndRenterEmailIgnoreCaseAndStatusIn(
                car.getId(),
                currentUser.email(),
                List.of(BookingStatus.ACTIVE, BookingStatus.IN_TRIP)
            ).stream().findAny().isPresent();
        }
        
        CarResponseDTO dto = new CarResponseDTO(car, hasActiveBooking, currentUserId);
        
        // Populate owner stats for detailed view
        if (car.getOwner() != null) {
            Long ownerId = car.getOwner().getId();
            
            // Fetch average rating
            Double rating = reviewRepo.findAverageRatingForRevieweeAndDirection(
                    ownerId, ReviewDirection.FROM_USER
            );
            dto.setOwnerRating(rating != null ? rating : 0.0);
            
            // Fetch trip count
            long trips = bookingRepo.countByOwnerIdAndStatus(ownerId, BookingStatus.COMPLETED);
            dto.setOwnerTripCount((int) trips);
        }
        
        return dto;
    }

    @Transactional
    public CarResponseDTO updateCar(Long carId, CarRequestDTO dto, User requester) {
        Car car = repo.findById(carId)
                .orElseThrow(() -> new RuntimeException("Car not found with ID: " + carId));

        // Verify ownership
        if (!car.getOwner().getId().equals(requester.getId())) {
            throw new RuntimeException("You do not have permission to edit this car");
        }

        // Update fields
        if (dto.getBrand() != null && !dto.getBrand().isBlank()) {
            car.setBrand(dto.getBrand().trim());
        }
        if (dto.getModel() != null && !dto.getModel().isBlank()) {
            car.setModel(dto.getModel().trim());
        }
        if (dto.getYear() != null) {
            car.setYear(dto.getYear());
        }
        // BigDecimal price validation (minimum 10 RSD)
        if (dto.getPricePerDay() != null && dto.getPricePerDay().compareTo(BigDecimal.TEN) >= 0) {
            car.setPricePerDay(dto.getPricePerDay());
        }
        if (dto.getLocation() != null && !dto.getLocation().isBlank()) {
            car.setLocation(dto.getLocation().trim().toLowerCase());
        }
        if (dto.getLicensePlate() != null && !dto.getLicensePlate().isBlank()) {
            car.setLicensePlate(dto.getLicensePlate().trim());
        }
        if (dto.getDescription() != null) {
            car.setDescription(dto.getDescription());
        }
        if (dto.getSeats() != null) {
            car.setSeats(dto.getSeats());
        }
        if (dto.getFuelType() != null) {
            car.setFuelType(dto.getFuelType());
        }
        if (dto.getFuelConsumption() != null) {
            car.setFuelConsumption(dto.getFuelConsumption());
        }
        if (dto.getTransmissionType() != null) {
            car.setTransmissionType(dto.getTransmissionType());
        }
        if (dto.getFeatures() != null) {
            car.getFeatures().clear();
            car.getFeatures().addAll(dto.getFeatures());
        }
        if (dto.getAddOns() != null) {
            car.getAddOns().clear();
            car.getAddOns().addAll(dto.getAddOns());
        }
        if (dto.getCancellationPolicy() != null) {
            car.setCancellationPolicy(dto.getCancellationPolicy());
        }
        if (dto.getMinRentalDays() != null) {
            car.setMinRentalDays(dto.getMinRentalDays());
        }
        if (dto.getMaxRentalDays() != null) {
            car.setMaxRentalDays(dto.getMaxRentalDays());
        }
        if (dto.getImageUrl() != null) {
            car.setImageUrl(dto.getImageUrl());
        }
        if (dto.getImageUrls() != null) {
            // Use setImageUrls to properly manage CarImage entities
            car.setImageUrls(new ArrayList<>(dto.getImageUrls()));
        }

        Car savedCar = repo.save(car);
        // Return DTO with exact location for owner
        return new CarResponseDTO(savedCar, true, requester.getId());
    }

    @Transactional
    public CarResponseDTO toggleAvailability(Long carId, boolean available, User requester) {
        Car car = repo.findById(carId)
                .orElseThrow(() -> new RuntimeException("Car not found with ID: " + carId));

        // Verify ownership
        if (!car.getOwner().getId().equals(requester.getId())) {
            throw new RuntimeException("You do not have permission to modify this car");
        }

        // CRITICAL: Prevent activation if car is not approved
        if (available && car.getApprovalStatus() != ApprovalStatus.APPROVED) {
            throw new RuntimeException("Cannot activate car that is not approved by admin. Current status: " + car.getApprovalStatus());
        }

        car.setAvailable(available);
        Car savedCar = repo.save(car);
        // Return DTO with exact location for owner
        return new CarResponseDTO(savedCar, true, requester.getId());
    }

    /**
     * Delete car method - DEPRECATED
     * @deprecated Car deletion is disabled. Use toggleAvailability() instead to deactivate cars.
     * This method will be removed in a future release.
     * @throws RuntimeException always, as deletion is no longer supported
     */
    @Deprecated
    @Transactional
    public void deleteCar(Long carId, User requester) {
        // Safety check: prevent deletion
        throw new RuntimeException("Car deletion is disabled. Use toggleAvailability() to deactivate cars instead.");
    }

    /**
     * Search cars with dynamic filtering, sorting, and pagination
     * @param criteria Search criteria with optional filters
     * @return Page of car response DTOs
     */
    @Transactional(readOnly = true)
    public Page<CarResponseDTO> searchCars(CarSearchCriteria criteria) {
        // Normalize and validate criteria
        criteria.normalize();

        // Build pageable with sorting
        Pageable pageable = buildPageable(criteria);
        
        // CRITICAL: Only show APPROVED cars in public search
        // Use native query to work around PostgreSQL ENUM operator issue
        // Native SQL: approval_status::text = 'APPROVED' handles the ENUM properly
        Page<Car> carPage = repo.findApprovedAvailableCars(pageable);

        // Map to DTOs with privacy-aware location
        Long currentUserId = currentUser.idOrNull();
        return carPage.map(car -> mapToResponseWithPrivacy(car, currentUserId));
    }

    /**
     * Build Pageable object with sorting from criteria
     */
    private Pageable buildPageable(CarSearchCriteria criteria) {
        int page = criteria.getPage() != null ? criteria.getPage() : 0;
        int size = criteria.getSize() != null ? criteria.getSize() : 20;

        // Parse sort parameter (e.g., "price,asc" or "year,desc")
        Sort sort = Sort.unsorted();
        if (criteria.getSort() != null && !criteria.getSort().isBlank()) {
            String[] sortParts = criteria.getSort().split(",");
            if (sortParts.length == 2) {
                String field = sortParts[0].trim();
                String direction = sortParts[1].trim();

                // Whitelist allowed sort fields to prevent arbitrary sorting
                if (isValidSortField(field)) {
                    Sort.Direction sortDirection = direction.equalsIgnoreCase("desc")
                            ? Sort.Direction.DESC
                            : Sort.Direction.ASC;
                    sort = Sort.by(sortDirection, field);
                }
            }
        } else {
            // Default sorting: newest first (by id DESC as proxy for creation order)
            sort = Sort.by(Sort.Direction.DESC, "id");
        }

        return PageRequest.of(page, size, sort);
    }

    /**
     * Validate sort field to prevent arbitrary field sorting
     */
    private boolean isValidSortField(String field) {
        return field.equals("pricePerDay") ||
                field.equals("year") ||
                field.equals("brand") ||
                field.equals("model") ||
                field.equals("seats") ||
                field.equals("id");
    }

    /**
     * Map Car to CarResponseDTO with privacy-aware location handling.
     * 
     * Privacy Logic:
     * - Owner sees exact location (isOwner check in DTO constructor)
     * - Active booker sees exact location (passed via hasActiveBooking parameter)
     * - Others see fuzzy coordinates (±500m) and city-only address
     * 
     * @param car The car entity
     * @param currentUserId The current user's ID (null if anonymous)
     * @return CarResponseDTO with appropriate location privacy
     */
    private CarResponseDTO mapToResponseWithPrivacy(Car car, Long currentUserId) {
        // For list views, we don't check active bookings (performance)
        // Active booking check is done in getCarById for detail views
        return new CarResponseDTO(car, false, currentUserId);
    }

    private CarResponseDTO mapToResponse(Car car) {
        return new CarResponseDTO(car);
    }

    /**
     * Get all distinct car makes from the database.
     * Used for filter dropdowns.
     * 
     * P0-5 FIX: Uses optimized DISTINCT query instead of loading all cars.
     */
    public List<String> getAllMakes() {
        return repo.findDistinctBrands();
    }
}

File: rentoza-frontend/src/app/core/services/car.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable, map } from 'rxjs';

import { environment } from '@environments/environment';
import { Car } from '@core/models/car.model';
import { Review } from '@core/models/review.model';
import { CarSearchCriteria, PagedResponse } from '@core/models/car-search.model';
import { isWithinRadius } from '@core/utils/distance.util';
import { normalizeSearchString } from '@core/utils/string-normalization.util';

/**
 * Backend Car DTO interface (uses 'brand' instead of 'make')
 */
interface BackendCarDTO {
  id: string;
  brand: string; // Backend uses 'brand'
  model: string;
  year: number;
  pricePerDay: number;
  location: string;
  description?: string;
  imageUrl?: string;
  available?: boolean;
  seats?: number;
  fuelType?: string;
  fuelConsumption?: number;
  transmissionType?: string;
  features?: string[];
  addOns?: string[];
  cancellationPolicy?: string;
  minRentalDays?: number;
  maxRentalDays?: number;
  imageUrls?: string[];
  ownerFullName?: string;
  ownerEmail?: string;
}

@Injectable({ providedIn: 'root' })
export class CarService {
  private readonly baseUrl = `${environment.baseApiUrl}/cars`;

  constructor(private readonly http: HttpClient) {}

  /**
   * Map backend DTO to frontend Car model
   * Backend uses 'brand', frontend uses 'make'
   */
  private mapBackendCarToFrontend(backendCar: any): Car {
    const { brand, ...rest } = backendCar;
    return {
      ...rest,
      make: brand, // Map 'brand' to 'make'
    } as Car;
  }

  getCars(): Observable<Car[]> {
    return this.http
      .get<any[]>(this.baseUrl)
      .pipe(map((cars) => cars.map((car) => this.mapBackendCarToFrontend(car))));
  }

  /**
   * Search cars with filters, sorting, and pagination
   * @param criteria Search criteria
   * @returns Paginated response with cars
   */
  searchCars(criteria: CarSearchCriteria): Observable<PagedResponse<Car>> {
    let params = new HttpParams();

    // Add all non-null/undefined criteria to params
    // Apply normalization to text-based filters
    if (criteria.minPrice !== undefined) {
      params = params.set('minPrice', criteria.minPrice.toString());
    }
    if (criteria.maxPrice !== undefined) {
      params = params.set('maxPrice', criteria.maxPrice.toString());
    }
    if (criteria.vehicleType) {
      params = params.set('vehicleType', normalizeSearchString(criteria.vehicleType));
    }
    if (criteria.make) {
      params = params.set('make', normalizeSearchString(criteria.make));
    }
    if (criteria.model) {
      params = params.set('model', normalizeSearchString(criteria.model));
    }
    if (criteria.minYear !== undefined) {
      params = params.set('minYear', criteria.minYear.toString());
    }
    if (criteria.maxYear !== undefined) {
      params = params.set('maxYear', criteria.maxYear.toString());
    }
    if (criteria.location) {
      params = params.set('location', normalizeSearchString(criteria.location));
    }
    if (criteria.minSeats !== undefined) {
      params = params.set('minSeats', criteria.minSeats.toString());
    }
    if (criteria.transmission) {
      params = params.set('transmission', criteria.transmission);
    }
    if (criteria.features && criteria.features.length > 0) {
      // Send features as comma-separated string
      params = params.set('features', criteria.features.join(','));
    }
    if (criteria.page !== undefined) {
      params = params.set('page', criteria.page.toString());
    }
    if (criteria.size !== undefined) {
      params = params.set('size', criteria.size.toString());
    }
    if (criteria.sort) {
      params = params.set('sort', criteria.sort);
    }

    return this.http.get<any>(`${this.baseUrl}/search`, { params }).pipe(
      map((response) => ({
        content: response.content.map((car: any) => this.mapBackendCarToFrontend(car)),
        totalElements: response.totalElements,
        totalPages: response.totalPages,
        currentPage: response.currentPage,
        pageSize: response.pageSize,
        hasNext: response.hasNext,
        hasPrevious: response.hasPrevious,
      }))
    );
  }

  /**
   * Get the maximum price per day from all available cars
   * Used for dynamic price slider range
   */
  getMaxPrice(): Observable<number> {
    const params = new HttpParams()
      .set('page', '0')
      .set('size', '1')
      .set('sort', 'pricePerDay,desc');

    return this.http.get<any>(`${this.baseUrl}/search`, { params }).pipe(
      map((response) => {
        if (response.content && response.content.length > 0) {
          return Math.ceil(response.content[0].pricePerDay);
        }
        return 500; // Fallback to 500 if no cars found
      })
    );
  }

  /**
   * Get cars by location with 20km radius filtering
   * @param location Search location (city name)
   * @param radiusKm Search radius in kilometers (default: 20)
   */
  getCarsByLocation(location: string, radiusKm: number = 20): Observable<Car[]> {
    // Get all cars and filter client-side for radius search
    return this.getCars().pipe(
      map((cars) => cars.filter((car) => isWithinRadius(car.location, location, radiusKm)))
    );
  }

  /**
   * Alternative: Use backend endpoint if it exists and supports radius search
   * This is a fallback method in case backend implements radius filtering
   */
  getCarsByLocationBackend(location: string): Observable<Car[]> {
    return this.http
      .get<any[]>(`${this.baseUrl}/location/${location}`)
      .pipe(map((cars) => cars.map((car) => this.mapBackendCarToFrontend(car))));
  }

  getCarById(id: string): Observable<Car> {
    return this.http
      .get<any>(`${this.baseUrl}/${id}`)
      .pipe(map((car) => this.mapBackendCarToFrontend(car)));
  }

  getCarReviews(id: string): Observable<Review[]> {
    return this.http.get<Review[]>(`${this.baseUrl}/${id}/reviews`);
  }

  // ========== Owner Operations ==========

  /**
   * Get all cars for a specific owner
   * @param ownerEmail Email of the owner
   */
  getOwnerCars(ownerEmail: string): Observable<Car[]> {
    return this.http
      .get<any[]>(`${this.baseUrl}/owner/${ownerEmail}`)
      .pipe(map((cars) => cars.map((car) => this.mapBackendCarToFrontend(car))));
  }

  /**
   * Add a new car (owner only)
   * @param carData Car data to add (frontend format with 'make')
   */
  addCar(carData: Partial<Car>): Observable<Car> {
    // Map frontend 'make' to backend 'brand'
    const { make, ...rest } = carData;
    const backendData = {
      ...rest,
      brand: make,
    };

    return this.http
      .post<any>(`${this.baseUrl}/add`, backendData)
      .pipe(map((car) => this.mapBackendCarToFrontend(car)));
  }

  /**
   * Update an existing car (owner only)
   * @param id Car ID
   * @param carData Updated car data (frontend format with 'make')
   */
  updateCar(id: string, carData: Partial<Car>): Observable<Car> {
    // Map frontend 'make' to backend 'brand'
    const { make, ...rest } = carData;
    const backendData = {
      ...rest,
      brand: make,
    };

    return this.http
      .put<any>(`${this.baseUrl}/${id}`, backendData)
      .pipe(map((car) => this.mapBackendCarToFrontend(car)));
  }

  /**
   * Delete a car (owner only)
   * @param id Car ID
   */
  deleteCar(id: string): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }

  /**
   * Toggle car availability (owner only)
   * @param id Car ID
   * @param available New availability status
   */
  toggleAvailability(id: string, available: boolean): Observable<Car> {
    return this.http
      .patch<any>(`${this.baseUrl}/${id}/availability`, { available })
      .pipe(map((car) => this.mapBackendCarToFrontend(car)));
  }
}
